Simple Javascript Clock (Workshop Bounty) (#1300)
hackclub/hackclub: workshops/Simple Js Clock/README.md

---
name: 'JavaScript Clock'
description: 'Build a clock using JavaScript.'
author: '@FaisalSayed'
---

# JavaScript Clock  

Have you ever wondered how to build a clock? Have you ever wanted to build your own clock but didn't because it was hard?  
Well today I'm going to show you how to build one and you know what? It's very simple!  

It will look something like this (But make sure you do add some custom CSS!) :  
![Image](https://cloud-azpvjrywa.vercel.app/image.png)

Here's the [live demo][final_live_demo] and [final code][final_code].

[final_live_demo]: https://js-css-clock.faisalsayed1.repl.co/
[final_code]: https://repl.it/@FaisalSayed1/JS-CSS-Clock
[repl_it]: https://repl.it

> ## Part I: Prerequisites

Reviewer: Change part numbers to normal numbers instead of Roman Numerals.

You should have a beginner understanding of:
- HTML
- CSS
- JavaScript

## Part II: Setup

### Setting up your code environment on Repl.it

> [Repl.it](https://repl.it) is an online code editor where you will build your clock today.

Reviewer: ```suggestion
[Repl.it](https://repl.it) is an online code editor where you can build your clock today. You don't have to use Repl.it but we suggest you do as it sets everything up for you.
```

To get started, go to [https://repl.it/languages/html](https://repl.it/languages/html). Your coding environment will spin up in just a few seconds!

This is what it looks like:  
![Image](https://cloud-me8mfihcw.vercel.app/image.png)  
## Part III: Inspecting The Default Files

![Image](https://media.giphy.com/media/3o7bufkPz3LRof205G/giphy.gif)

### 1) The HTML file
Let us first take a look at the HTML file (`index.html`) present in your repl.

At line 1 we have `<!DOCTYPE html>` This declares that this file is a HTML file.  
If we take a look in the `<html>` tag, we will find a `<body>` tag. Here's where we will write the code. Make sure that all the code you write is included between the opening and closing body tags. (`<body></body>`). Make sure to remove line 5 (`<meta name="viewport" content="width=device-width">`) which may cause some weird responsiveness issues for our workshop.  

![Image](https://cloud-9zj6930c6.vercel.app/gif_maker.gif)

### 2) The CSS file

> This is the `style.css` file in your directory. Whatever styles we give to our website is written in this file. If you take a look in the `<head>` tag in your html, you will find a line of code (`<link href="style.css" rel="stylesheet" type="text/css" />`). This means that your CSS file is properly linked to your HTML file!

Reviewer: ```suggestion
This is the `style.css` file in your directory. Whatever styles we give to our website is written in this file. If you take a look in the `<head>` tag in your HTML, you will find a line of code (`<link href="style.css" rel="stylesheet" type="text/css" />`). This means that your HTML file is properly linked to your CSS file!
```

### 3) The JavaScript file

This is the `script.js` file in your directory. It's empty for now, but we will fill it eventually through the workshop!  
Also, this is the most important file for our today's workshop. The code written here will be responsible for the functioning of our clock.  
If you take a look in the `<body>` tag, you'll find:

```html
<script src="script.js"></script>
```

> This means that your JavaScript file is linked to your HTML.

Reviewer: ```suggestion
This means that your HTML file is linked to your JavaScript.
```

## Part IV: Building the Clock

### 1) HTML

Lets start building our clock! First let's write the HTML code required.  
Inside of our `<body>` tag, add a `<div>` tag with a class "clock".  

Add another `<div>` tag inside of your existing `<div>` and give it a class of "clock-face". Let's add 3 more `<div>` tags nested inside of the "clock-face" div. 

Give a class of `hand hour-hand`, `hand min-hand`, `hand second-hand` to each of the 3 new `<div>` tags respectively.  

Note: Make sure the `<script>` tag, which is in your body, is below the `<div>` tags or else it will throw you errors.  

> This is your code so far:

Reviewer: ```suggestion
This is what your code should look like so far:
```

2. Open curly brackets `{}`.
3. Start adding your styles inside the curly brackets!

We start by giving our website a soothing green color. Go ahead and try changing it to the color of your own choice!  

Then we removed the default margin and padding from all the elements in our website (Some elements have a default value for their margin and padding). We changed the default display `(display: block)` to `(display: flex)`.  

A very good reason for this is because `flex` enables you to align your items perfectly to the center! And that's what we have done in the next two lines. The `align-items` property aligns all your children elements to center, but horizontally. The `justify-content` property aligns all your children elements to center, but vertically.  

Next we have the `min-height` property which sets a minimum height of `100vh` (viewport height) to your `<body>`. The `min-height` property works together with the `align-items` property and helps the children to align properly at the center.

Learn more about [viewport heights](https://www.sitepoint.com/css-viewport-units-quick-start/)

Next, we'll need to add some styles to our clock!  

> Following the 3 main steps given above, let's accomplish this task!  
> 1. Select the class designated to the clock div.  

Reviewer: ```suggestion
Following the 3 main steps given above, let's accomplish this task!
  
1. Select the class designated to the clock div.  
```

First let us link the `<div>` tags in our HTML to JavaScript:  
Type the following code:  

```js
const secondHand = document.querySelector(".second-hand");
const minsHand = document.querySelector(".min-hand");
const hourHand = document.querySelector(".hour-hand");
```

**Explanation:**
We declare the variables `secondHand`, `minsHand`, `hourHand` using 'const' and set It's value to the first element that matches a specified CSS class(s) in the document. In our case, that CSS selectors are `.second-hand` , `.min-hand` , `.hour-hand` respectively.  

**What is querySelector()?**
The `querySelector()` method returns the first element that matches a specified CSS class(s) in the document.

Learn more about `query selector` [here](https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelector).

Now It's time to implement a function which will help us in the proper functioning of the clock.  
Create a function `setDate` in your file. (You can name it anything you want)

```js
function setDate() {
}
```

We can get the current date using the `JavaScript Date Object`. Let's store it in a variable:

```js
function setDate() {
 	const currentTime = new Date();
}
```

> This object contains specific methods through which we can get the current hour / minute / second.  

Reviewer: ```suggestion
The `Date` object contains specific methods through which we can get the current hour / minute / second.  
```

Learn more on [JavaScript dates](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date).

lets start by getting the current seconds, Add this code to the `setDate` function:  
```
const seconds = currentTime.getSeconds();
```
Now, if we want the hands to rotate, we can't simply feed the current time to it. We'll need to convert the current values into degrees and then do the transition.  

Time to do some maths! 

![Image](https://media.giphy.com/media/26gR0YFZxWbnUPtMA/giphy.gif)

Code:

```js
function setDate() {
	const currentTime = new Date();
	const seconds = currentTime.getSeconds();
	const secondDegrees = ((seconds / 60) * 360) + 90;
	secondHand.style.transform = `rotate(${secondDegrees}deg)`;
}
```

> **Explanation:**

Reviewer: ```suggestion

**Explanation:**
```
